### ISSUE TITLE
perf: Selective Gradient Checkpointing with Policy-Based Rematerialization

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#83]
## Summary

Implement selective gradient checkpointing that trades compute for memory by recomputing activations during backward pass based on configurable policies.

**Inspiration:** JAX `jax/_src/ad_checkpoint.py:62-150` - provides checkpoint policies like `everything_saveable`, `dots_saveable`, and `offload_dot_with_no_batch_dims`.

## Problem

Training large models exhausts GPU memory storing activations for backward pass:
- Transformer with 1B params: ~16GB just for activations at batch=32
- Current options: reduce batch size (slower) or gradient accumulation (complex)
- No way to selectively trade compute for memory

### 1. Checkpoint Policy Trait

```rust
use crate::autograd::{Operation, OperationType, TensorId};

/// Policy for deciding which activations to save vs recompute
pub trait CheckpointPolicy: Send + Sync {
    /// Returns true if this operation's output should be saved
    fn should_save(&self, op: &Operation) -> bool;
    
    /// Optional: estimate recomputation cost (for optimization)
    fn recompute_cost(&self, op: &Operation) -> f64 {
        1.0  // Default: all ops have equal cost
    }
    
    /// Optional: memory cost of saving this activation
    fn memory_cost(&self, op: &Operation) -> usize {
        op.output_size_bytes()
    }
}

/// Built-in policies inspired by JAX
pub mod policies {
    use super::*;
    
    /// Save everything - maximum memory, minimum compute
    pub struct SaveAll;
    
    impl CheckpointPolicy for SaveAll {
        fn should_save(&self, _op: &Operation) -> bool {
            true
        }
    }
    
    /// Save nothing - minimum memory, maximum compute (full recomputation)
    pub struct SaveNothing;
    
    impl CheckpointPolicy for SaveNothing {
        fn should_save(&self, _op: &Operation) -> bool {
            false
        }
    }
    
    /// Save only matrix multiplications (expensive to recompute)
    pub struct SaveMatmuls;
    
    impl CheckpointPolicy for SaveMatmuls {
        fn should_save(&self, op: &Operation) -> bool {
            matches!(op.op_type, 
                OperationType::Matmul | 
                OperationType::BatchedMatmul |
                OperationType::Conv2d |
                OperationType::Linear
            )
        }
        
        fn recompute_cost(&self, op: &Operation) -> f64 {
            match op.op_type {
                OperationType::Matmul => 100.0,
                OperationType::BatchedMatmul => 150.0,
                OperationType::Conv2d => 200.0,
                _ => 1.0,
            }
        }
    }
    
    /// Save matmuls without batch dimensions (common in transformers)
    /// These are the most expensive and benefit most from caching
    pub struct SaveUnbatchedMatmuls;
    
    impl CheckpointPolicy for SaveUnbatchedMatmuls {
        fn should_save(&self, op: &Operation) -> bool {
            match op.op_type {
                OperationType::Matmul | OperationType::Linear => {
                    // Check if any input has batch dim
                    !op.inputs.iter().any(|t| t.shape().len() > 2)
                }
                _ => false,
            }
        }
    }
    
    /// Offload to CPU instead of recomputing
    pub struct OffloadToCpu {
        /// Maximum CPU memory to use for offloading
        pub max_cpu_bytes: usize,
        /// Current CPU memory used
        used_bytes: std::sync::atomic::AtomicUsize,
    }
    
    impl CheckpointPolicy for OffloadToCpu {
        fn should_save(&self, op: &Operation) -> bool {
            // Save to CPU if we have room
            let size = op.output_size_bytes();
            let current = self.used_bytes.load(std::sync::atomic::Ordering::Relaxed);
            
            if current + size <= self.max_cpu_bytes {
                self.used_bytes.fetch_add(size, std::sync::atomic::Ordering::Relaxed);
                true
            } else {
                false
            }
        }
    }
    
    /// Custom predicate-based policy
    pub struct Custom<F: Fn(&Operation) -> bool + Send + Sync> {
        predicate: F,
    }
    
    impl<F: Fn(&Operation) -> bool + Send + Sync> CheckpointPolicy for Custom<F> {
        fn should_save(&self, op: &Operation) -> bool {
            (self.predicate)(op)
        }
    }
}
```

### 2. Checkpointed Computation Graph

```rust
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

/// A computation graph with selective checkpointing
pub struct CheckpointedGraph {
    /// All operations in topological order
    operations: Vec<Operation>,
    /// Which operations have saved activations
    saved_activations: HashMap<TensorId, Arc<Tensor>>,
    /// Policy for deciding what to save
    policy: Box<dyn CheckpointPolicy>,
    /// Dependency graph for recomputation
    dependencies: HashMap<TensorId, Vec<TensorId>>,
}

impl CheckpointedGraph {
    pub fn new(policy: impl CheckpointPolicy + 'static) -> Self {
        Self {
            operations: Vec::new(),
            saved_activations: HashMap::new(),
            policy: Box::new(policy),
            dependencies: HashMap::new(),
        }
    }
    
    /// Record a forward operation
    pub fn record_forward(&mut self, op: Operation, output: Arc<Tensor>) -> TensorId {
        let output_id = output.id();
        
        // Record dependencies
        for input in &op.inputs {
            self.dependencies
                .entry(output_id)
                .or_default()
                .push(input.id());
        }
        
        // Decide whether to save
        if self.policy.should_save(&op) {
            self.saved_activations.insert(output_id, output);
        }
        
        self.operations.push(op);
        output_id
    }
    
    /// Get activation, recomputing if necessary
    pub fn get_activation(&mut self, id: TensorId) -> Arc<Tensor> {
        // Fast path: already saved
        if let Some(tensor) = self.saved_activations.get(&id) {
            return tensor.clone();
        }
        
        // Slow path: recompute
        self.recompute(id)
    }
    
    /// Recompute an activation from saved checkpoints
    fn recompute(&mut self, target_id: TensorId) -> Arc<Tensor> {
        // Find path from nearest checkpoint to target
        let path = self.find_recompute_path(target_id);
        
        // Execute operations in order
        let mut intermediates: HashMap<TensorId, Arc<Tensor>> = HashMap::new();
        
        for op in path {
            // Gather inputs (from saved or just-computed)
            let inputs: Vec<Arc<Tensor>> = op.input_ids.iter()
                .map(|&id| {
                    self.saved_activations.get(&id)
                        .or_else(|| intermediates.get(&id))
                        .cloned()
                        .expect("Missing input for recomputation")
                })
                .collect();
            
            // Recompute
            let output = op.execute(&inputs);
            intermediates.insert(op.output_id, output);
        }
        
        intermediates.remove(&target_id).expect("Target not computed")
    }
    
    /// BFS to find shortest path from checkpoint to target
    fn find_recompute_path(&self, target_id: TensorId) -> Vec<&Operation> {
        // Find all saved tensors
        let saved_ids: HashSet<TensorId> = self.saved_activations.keys().copied().collect();
        
        // BFS backward from target to find closest checkpoint
        let mut to_compute = Vec::new();
        let mut queue = vec![target_id];
        let mut visited = HashSet::new();
        
        while let Some(id) = queue.pop() {
            if saved_ids.contains(&id) || visited.contains(&id) {
                continue;
            }
            visited.insert(id);
            
            // Find operation that produced this tensor
            if let Some(op) = self.operations.iter().find(|o| o.output_id == id) {
                to_compute.push(op);
                for &input_id in &op.input_ids {
                    queue.push(input_id);
                }
            }
        }
        
        // Reverse to get topological order
        to_compute.reverse();
        to_compute
    }
}
```

### 3. Gradient Tape with Checkpointing

```rust
/// Gradient tape that supports checkpointing
pub struct CheckpointedTape {
    graph: CheckpointedGraph,
    /// Gradients accumulated during backward
    gradients: HashMap<TensorId, Tensor>,
}

impl CheckpointedTape {
    pub fn new(policy: impl CheckpointPolicy + 'static) -> Self {
        Self {
            graph: CheckpointedGraph::new(policy),
            gradients: HashMap::new(),
        }
    }
    
    /// Backward pass with selective recomputation
    pub fn backward(&mut self, loss: &Tensor) {
        // Initialize gradient of loss as 1
        self.gradients.insert(loss.id(), Tensor::ones_like(loss));
        
        // Process operations in reverse order
        for op in self.graph.operations.iter().rev() {
            // Get output gradient
            let output_grad = match self.gradients.get(&op.output_id) {
                Some(g) => g.clone(),
                None => continue,  // Not on path to loss
            };
            
            // Get activations (potentially recomputing)
            let activations: Vec<Arc<Tensor>> = op.input_ids.iter()
                .map(|&id| self.graph.get_activation(id))
                .collect();
            
            // Compute input gradients using chain rule
            let input_grads = op.backward(&activations, &output_grad);
            
            // Accumulate gradients
            for (input_id, grad) in op.input_ids.iter().zip(input_grads) {
                self.gradients
                    .entry(*input_id)
                    .and_modify(|g| *g = &*g + &grad)
                    .or_insert(grad);
            }
        }
    }
    
    /// Get gradient for a tensor
    pub fn gradient(&self, tensor: &Tensor) -> Option<&Tensor> {
        self.gradients.get(&tensor.id())
    }
}
```

### 4. High-Level API with Decorator Pattern

```rust
/// Checkpoint a function's activations
pub fn checkpoint<F, T>(func: F, policy: impl CheckpointPolicy + 'static) -> impl Fn(T) -> Tensor
where
    F: Fn(T) -> Tensor,
    T: Clone,
{
    move |input: T| {
        // Create checkpointed tape for this call
        let mut tape = CheckpointedTape::new(/* policy clone */);
        
        // Execute with recording
        let output = func(input);
        
        // Return with tape attached
        output
    }
}

/// Macro for easy checkpointing
#[macro_export]
macro_rules! checkpoint {
    // Checkpoint with SaveMatmuls policy (most common)
    ($func:expr) => {
        checkpoint($func, policies::SaveMatmuls)
    };
    
    // Checkpoint with specific policy
    ($func:expr, $policy:expr) => {
        checkpoint($func, $policy)
    };
}

// Usage:
// let checkpointed_transformer = checkpoint!(|x| transformer_block(x));
```

### 5. Memory-Optimal Checkpointing (Binomial Strategy)

```rust
/// Compute optimal checkpoint schedule for sequential layers
/// Based on: "Training Deep Nets with Sublinear Memory Cost" (Chen et al.)
pub struct BinomialCheckpointing {
    /// Total number of layers
    num_layers: usize,
    /// Available memory slots for checkpoints
    memory_slots: usize,
}

impl BinomialCheckpointing {
    /// Compute which layers to checkpoint
    pub fn checkpoint_indices(&self) -> Vec<usize> {
        // Use sqrt(n) checkpoints for O(sqrt(n)) memory
        let num_checkpoints = (self.num_layers as f64).sqrt().ceil() as usize;
        let interval = self.num_layers / num_checkpoints;
        
        (0..self.num_layers)
            .step_by(interval.max(1))
            .collect()
    }
}

/// Apply binomial checkpointing to sequential model
pub fn checkpoint_sequential<T>(
    layers: &[Box<dyn Layer<T>>],
    input: Tensor,
    memory_budget: usize,
) -> Tensor {
    let strategy = BinomialCheckpointing {
        num_layers: layers.len(),
        memory_slots: memory_budget / std::mem::size_of::<Tensor>(),
    };
    
    let checkpoint_indices: HashSet<usize> = strategy.checkpoint_indices().into_iter().collect();
    let mut saved = HashMap::new();
    
    let mut x = input;
    for (i, layer) in layers.iter().enumerate() {
        x = layer.forward(&x);
        if checkpoint_indices.contains(&i) {
            saved.insert(i, x.clone());
        }
    }
    
    x
}
```

### 6. CPU Offloading Implementation

```rust
use std::sync::mpsc::{channel, Sender, Receiver};
use std::thread;

/// Async CPU offloader for activations
pub struct CpuOffloader {
    /// Channel to send tensors for offloading
    tx: Sender<OffloadRequest>,
    /// Storage for offloaded tensors
    cpu_storage: Arc<RwLock<HashMap<TensorId, Vec<u8>>>>,
}

enum OffloadRequest {
    Store { id: TensorId, data: Tensor },
    Load { id: TensorId, response: oneshot::Sender<Tensor> },
}

impl CpuOffloader {
    pub fn new() -> Self {
        let (tx, rx) = channel();
        let cpu_storage = Arc::new(RwLock::new(HashMap::new()));
        
        // Spawn background thread for async transfers
        let storage_clone = cpu_storage.clone();
        thread::spawn(move || {
            Self::background_worker(rx, storage_clone);
        });
        
        Self { tx, cpu_storage }
    }
    
    /// Offload tensor to CPU (async, returns immediately)
    pub fn offload(&self, tensor: &Tensor) {
        let _ = self.tx.send(OffloadRequest::Store {
            id: tensor.id(),
            data: tensor.clone(),
        });
    }
    
    /// Reload tensor from CPU (blocks until ready)
    pub fn reload(&self, id: TensorId) -> Tensor {
        let (response_tx, response_rx) = oneshot::channel();
        let _ = self.tx.send(OffloadRequest::Load { id, response: response_tx });
        response_rx.recv().expect("Offloader died")
    }
    
    fn background_worker(
        rx: Receiver<OffloadRequest>,
        storage: Arc<RwLock<HashMap<TensorId, Vec<u8>>>>,
    ) {
        while let Ok(req) = rx.recv() {
            match req {
                OffloadRequest::Store { id, data } => {
                    // Copy to CPU memory (pinned if available)
                    let bytes = data.to_bytes();
                    storage.write().unwrap().insert(id, bytes);
                }
                OffloadRequest::Load { id, response } => {
                    let bytes = storage.read().unwrap().get(&id).cloned();
                    if let Some(bytes) = bytes {
                        let tensor = Tensor::from_bytes(&bytes);
                        let _ = response.send(tensor);
                    }
                }
            }
        }
    }
}
```

## Acceptance Criteria

- [ ] `CheckpointPolicy` trait with 5+ built-in policies
- [ ] `CheckpointedGraph` with recomputation logic
- [ ] `CheckpointedTape` for autograd integration
- [ ] Binomial checkpointing for sequential models
- [ ] CPU offloading with async transfers
- [ ] Memory usage tracking and reporting
- [ ] Benchmarks: memory vs compute tradeoff curves
- [ ] Unit tests for correctness (gradients match non-checkpointed)

## Expected Performance Impact

| Model | Batch | No Checkpoint | SaveMatmuls | SaveNothing |
|-------|-------|---------------|-------------|-------------|
| BERT-base | 32 | 8GB, 1.0x | 4GB, 1.2x | 2GB, 2.0x |
| GPT-2 | 16 | 12GB, 1.0x | 5GB, 1.3x | 2.5GB, 2.5x |
| ResNet-152 | 64 | 10GB, 1.0x | 6GB, 1.1x | 3GB, 1.8x |

Memory savings of 2-5x with compute overhead of 20-150%.

## References

- JAX checkpoint: `jax/_src/ad_checkpoint.py:62-150`
- "Training Deep Nets with Sublinear Memory Cost": https://arxiv.org/abs/1604.06174
- PyTorch checkpointing: https://pytorch.org/docs/stable/checkpoint.html

## Labels

`performance`, `memory`, `autograd`, `P1-high`

**Comment by noahgift** [comment]
Implemented in commit 5a88ade: CheckpointPolicy trait, built-in policies (SaveAll, SaveNothing, SaveMatmuls, SaveUnbatchedMatmuls, BinomialCheckpointing, MemoryBudget, CustomPolicy), PolicyCheckpointManager, estimate_policy_tradeoff().
